pkg/tbtc/dkg_test.go

```diff
@@ -160,6 +165,211 @@ func TestRegisterSigner(t *testing.T) {
+	}
+
+	const (
+		groupSize       = 5
+		groupQuorum     = 3
+		honestThreshold = 2
+	)
+
+	tecdsaDkgResult := &dkg.Result{
+		Group:           group.NewGroup(groupSize-honestThreshold, groupSize),
+		PrivateKeyShare: tecdsa.NewPrivateKeyShare(testData[0]),
+	}
+
+	groupPublicKey, err := tecdsaDkgResult.GroupPublicKeyBytes()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var tests = map[string]struct {
+		submitterMemberIndex     group.MemberIndex
+		controlledMembersIndexes []group.MemberIndex
+		resultValid              bool
+		expectedEvent            interface{}
+		expectedDkgState         DKGState
+	}{
+		"result approved by the submitter": {
+			submitterMemberIndex:     group.MemberIndex(2),
+			controlledMembersIndexes: []group.MemberIndex{1, 2, 3, 4, 5},
+			resultValid:              true,
+			expectedEvent: &DKGResultApprovedEvent{
+				ResultHash: sha3.Sum256(groupPublicKey),
+				Approver:   "",
+				// 16 is the next block after 15 blocks of the challenge period
+				BlockNumber: 16,
+			},
+			expectedDkgState: Idle,
+		},
+		"result approved by a non-submitter": {
+			submitterMemberIndex:     group.MemberIndex(1),
+			controlledMembersIndexes: []group.MemberIndex{2, 3, 4, 5},
+			resultValid:              true,
+			expectedEvent: &DKGResultApprovedEvent{
+				ResultHash: sha3.Sum256(groupPublicKey),
+				Approver:   "",
+				// 26 is the next block after 15 blocks of the challenge period,
+				// 5 blocks of the precedence period, and 5 blocks of the delay
+				// for member 2
+				BlockNumber: 26,
+			},
+			expectedDkgState: Idle,
+		},
+		"result challenged": {
+			submitterMemberIndex:     group.MemberIndex(2),
+			controlledMembersIndexes: []group.MemberIndex{1, 2, 3, 4, 5},
+			resultValid:              false,
+			expectedEvent: &DKGResultChallengedEvent{
+				ResultHash:  sha3.Sum256(groupPublicKey),
+				Challenger:  "",
+				Reason:      "",
+				BlockNumber: 0, // challenge is submitted immediately
+			},
+			expectedDkgState: AwaitingResult,
+		},
+	}
+
+	for testName, test := range tests {
+		t.Run(testName, func(t *testing.T) {
+			localChain := Connect(groupSize, groupQuorum, honestThreshold)
+
+			operatorAddress, err := localChain.operatorAddress()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			operatorID, err := localChain.GetOperatorID(operatorAddress)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			signatures := make(map[group.MemberIndex][]byte)
+			operatorsIDs := make(chain.OperatorIDs, groupSize)
+			operatorsAddresses := make(chain.Addresses, groupSize)
+
+			for memberIndex := uint8(1); memberIndex <= groupSize; memberIndex++ {
+				signatures[memberIndex] = []byte{memberIndex}
+
+				if slices.Contains(test.controlledMembersIndexes, memberIndex) {
+					operatorsIDs[memberIndex-1] = operatorID
+					operatorsAddresses[memberIndex-1] = operatorAddress
+				}
+			}
+
+			groupSelectionResult := &GroupSelectionResult{
+				OperatorsIDs:       operatorsIDs,
+				OperatorsAddresses: operatorsAddresses,
+			}
```

Comment:
I am lost about `controlledMembersIndexes` construction and this test as a whole.

First, I think we have inconsistency in test data preparation: `signatures` will have `groupSize` entries but `GroupSelectionResult` will only have `len(controlledMembersIndexes)` entries.

Second, the submitter member index in `executeDkgValidation` is obtained from `operatorID, err := de.operatorIDFn()` and then looking at `*GroupSelectionResult`. Why do we care about a collection of `controlledMembersIndexes`?

Reply:
The `controlledMembersIndexes` is for simulating a case when the given member is not controlled by the tested node so their action can be omitted. For example, we want to check the approval after the submitter's precedence period. We use member 1 as the submitter but we do not make it "controlled" (their index is not in `controlledMembersIndexes`). This way, member 1 does not approve so the result can be approved by member 2 after the precedence period.

Reply:
I am sorry but I still do not follow. In this test (`TestDkgExecutor_ExecuteDkgValidation`) we test the `dkgExecutor.executeDkgValidation` function. `dkgExecutor` is created with only one `operatorIDFn`, so it is controlling exactly one operator - the one fetched from the local chain: `operatorID, err := localChain.GetOperatorID(operatorAddress)`. `test.controlledMembersIndexes` are used only in one place: to construct `operatorsIDs` and `operatorsAddresses` that are later used to construct `GroupSelectionResult`:
```
			for memberIndex := uint8(1); memberIndex <= groupSize; memberIndex++ {
				signatures[memberIndex] = []byte{memberIndex}

				if slices.Contains(test.controlledMembersIndexes, memberIndex) {
					operatorsIDs[memberIndex-1] = operatorID
					operatorsAddresses[memberIndex-1] = operatorAddress
				}
			}

			groupSelectionResult := &GroupSelectionResult{
				OperatorsIDs:       operatorsIDs,
				OperatorsAddresses: operatorsAddresses,
			}
```

With this code, `GroupSelectionResult` contain only operators from the  `test.controlledMembersIndexes` while it should contain `groupSize` of operators - all selected to the given group.

Reply:
See 9536de6a